src/server/journal/streamer.cc

```diff
@@ -11,62 +11,178 @@
       });
 }
 
 void JournalStreamer::Cancel() {
-  Finalize();  // Finalize must be called before UnregisterOnChange because we first need to stop
-               // writing to buffer and notify the all the producers.
-               // Writing to journal holds mutex protecting change_cb_arr_, than the fiber can
-               // preemt when calling NotifyWritten and it will not run again till notified.
-               // UnregisterOnChange will try to lock the mutex therefor calling UnregisterOnChange
-               // before Finalize may cause deadlock.
+  VLOG(1) << "JournalStreamer::Cancel";
+  is_stopped_ = true;
+  waker_.notifyAll();
   journal_->UnregisterOnChange(journal_cb_id_);
+  WaitForInflightToComplete();
+}
+
+size_t JournalStreamer::GetTotalBufferCapacities() const {
+  return in_flight_bytes_ + pending_buf_.capacity();
+}
 
-  if (write_fb_.IsJoinable()) {
-    write_fb_.Join();
+constexpr size_t kFlushThreshold = 2_KB;
+
+void JournalStreamer::Write(std::string_view str) {
+  DCHECK(!str.empty());
+  DVLOG(2) << "Writing " << str.size() << " bytes";
+
+  // If we do not have any in flight requests we send the string right a way.
+  // We can not aggregate it since we do not know when the next update will follow.
+  size_t total_pending = pending_buf_.size() + str.size();
+  if (in_flight_bytes_ == 0 || total_pending > kFlushThreshold) {
+    // because of potential SOO with strings we allocate explicitly on heap
+    uint8_t* buf(new uint8_t[str.size()]);
+
+    // TODO: it is possible to remove these redundant copies if we adjust high level
+    // interfaces to pass reference-counted buffers.
+    memcpy(buf, str.data(), str.size());
+    in_flight_bytes_ += total_pending;
+    io::Bytes src(buf, str.size());
+
+    if (pending_buf_.empty()) {
+      dest_->AsyncWrite(src, [buf, this, len = str.size()](std::error_code ec) {
+        delete[] buf;
+        OnCompletion(ec, len);
+      });
+    } else {
+      iovec v[2];
+      v[0] = IoVec(pending_buf_);
+      v[1] = IoVec(src);
+
+      dest_->AsyncWrite(
+          v, 2,
+          [buf0 = std::move(pending_buf_), buf, this, len = total_pending](std::error_code ec) {
+            delete[] buf;
+            OnCompletion(ec, len);
+          });
+    }
+    return;
   }
+
+  DCHECK_GT(in_flight_bytes_, 0u);
+  DCHECK_LE(pending_buf_.size() + str.size(), kFlushThreshold);
+
+  // Aggregate
+  size_t tail = pending_buf_.size();
+  pending_buf_.resize(pending_buf_.size() + str.size());
+  memcpy(pending_buf_.data() + tail, str.data(), str.size());
 }
 
-void JournalStreamer::WriterFb(io::Sink* dest) {
-  if (auto ec = ConsumeIntoSink(dest); ec) {
+void JournalStreamer::OnCompletion(std::error_code ec, size_t len) {
+  DCHECK_GE(in_flight_bytes_, len);
+
+  DVLOG(2) << "Completing from " << in_flight_bytes_ << " to " << in_flight_bytes_ - len;
+  in_flight_bytes_ -= len;
+  if (ec && !is_stopped_) {
+    is_stopped_ = true;
     cntx_->ReportError(ec);
```

Comment:
report error should lead to cntx->cancelled() being true, so is_stopped is redundant?

Reply:
yeah, I was contemplating over it. I will remove